modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+    {
+        return cv::log(*this + cv::power(*this * *this - Quat<T>(1,0,0,0), 0.5));
+    }
+
+    template <typename T>
+    inline Quat<T> atanh(const Quat<T> &q)
+    {
+        return q.atanh();
+    }
+
+    template <typename T>
+    inline Quat<T> Quat<T>::atanh() const
+    {
+        Quat<T> ident(1, 0, 0, 0);
+        Quat<T> c1 = (ident + *this).log();
+        Quat<T> c2 = (ident - *this).log();
+        return 0.5 * (c1 - c2);
+    }
+
+    template <typename T>
+    inline Quat<T> asin(const Quat<T> &q)
+    {
+        return q.asin();
+    }
+
+    template <typename T>
+    inline Quat<T> Quat<T>::asin() const
+    {
+        Quat<T> v(0, x, y, z);
+        T vNorm = v.norm();
+        T k = vNorm < CV_QUAT_EPS ? 1 : vNorm;
+        return -v / k * (*this * v / k).asinh();
+    }
+
+    template <typename T>
+    inline Quat<T> acos(const Quat<T> &q)
+    {
+        return q.acos();
+    }
+
+    template <typename T>
+    inline Quat<T> Quat<T>::acos() const
+    {
+        Quat<T> v(0, x, y, z);
+        T vNorm = v.norm();
+        T k = vNorm < CV_QUAT_EPS ? 1 : vNorm;
+        return -v / k * acosh();
+    }
+
+    template <typename T>
+    inline Quat<T> atan(const Quat<T> &q)
+    {
+        return q.atan();
+    }
+
+    template <typename T>
+    inline Quat<T> Quat<T>::atan() const
+    {
+        Quat<T> v(0, x, y, z);
+        T vNorm = v.norm();
+        T k = vNorm < CV_QUAT_EPS ? 1 : vNorm;
+        return -v / k * (*this * v / k).atanh();
+    }
+
+    template <typename T>
+    inline T Quat<T>::getAngle(bool assumeUnit) const
+    {
+        if (assumeUnit)
+        {
+            return 2 * std::acos(w);
+        }
+        if (x * x + y * y + z * z < CV_QUAT_EPS || norm() < CV_QUAT_EPS )
+        {
+            CV_Error(Error::StsBadArg, "This quaternion does not represent a rotation");
+        }
+        return 2 * std::acos(w / norm());
+    }
+
+    template <typename T>
+    inline Vec<T, 3> Quat<T>::getAxis(bool assumeUnit) const
+    {
+        T angle = getAngle(assumeUnit);
+        if (assumeUnit)
+        {
+            return Vec<T, 3>{x, y, z} / std::sin(angle / 2);
+        }
+        return Vec<T, 3> {x, y, z} / (norm() * std::sin(angle / 2));
+    }
+
+    template <typename T>
+    cv::Mat Quat<T>::toRotMat4x4(bool assumeUnit) const
+    {
+        T a = w, b = x, c = y, d = z;
+        if (!assumeUnit)
+        {
+            Quat<T> qTemp = normalize();
+            a = qTemp.w;
+            b = qTemp.x;
+            c = qTemp.y;
+            d = qTemp.z;
+        }
+        cv::Matx<T, 4, 4> R{
+            1,                      0,                       0,                   0,
+            0,1 - 2 * (c * c + d * d), 2 * (b * c + a * d)    , 2 * (b * d - a * c),
+            0,2 * (b * c - a * d)    , 1 - 2 * (b * b + d * d), 2 * (c * d + a * b),
+            0,2 * (b * d + a * c)    , 2 * (c * d - a * b)    , 1 - 2 * (b * b + c * c)};
+        return cv::Mat(R).t();
```

Comment:
It is better to write a transposed variant and avoid `MatExpr` creation.

Reply:
Could I ask why we should avoid `MatExpr`?

Can `MatExpr` handle complex operations with higher efficiency?

Reply:
`MatExpr` is too heavy for such small matrices: it involves `Mat` copying several times, sub-expression creation and so on. `MatExpr` has more sense for complex Mat-based expressions.
For small matrices there is a `Matx` which is almost overhead-free.

Reply:
OK. Thanks for your answer